### DISCUSSION TITLE
[Bug]: podman pull in yocto kirkstone

### DISCUSSION THREAD

**Discussion by Razriel3** [containers/podman#16983] [General]
### Issue Description

running on ubuntu 20.04 as build machine.
podman install on this machine is 3.4.2
when running from regular shell podman pull succesful
when running from honister version podman pull succesful
when running same recipe from kirkston version it fails.

### Steps to reproduce the issue

Steps to reproduce the issue
1. running bitbake of recipe. the recipe is trying to pull image using podman.
2. podman pull postgres:14
3.

### Describe the results you received

Log data follows:
| DEBUG: Executing shell function do_pull_image
| cannot setresgid: Invalid argument
| panic: runtime error: invalid memory address or nil pointer dereference
| [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5632d0bb17a0]
|
| goroutine 1 [running]:
| github.com/containers/podman/vendor/github.com/containers/common/libimage.(*Runtime).Pull(0x0, 0x5632d1b3d720, 0xc0007b56e0, 0xc0001f9c90, 0xb, 0x0, 0xc000159a40, 0x0, 0x0, 0x0, ...)
| /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/containers/common/libimage/pull.go:117 +0x500
| github.com/containers/podman/pkg/domain/infra/abi.(*ImageEngine).Pull(0xc00012c3d0, 0x5632d1b3d720, 0xc0007b56e0, 0x7ffde5cbbe0c, 0xb, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
| /usr/src/packages/BUILD/src/github.com/containers/podman/pkg/domain/infra/abi/images.go:231 +0x1e5
| github.com/containers/podman/cmd/podman/images.imagePull(0x5632d26d5e20, 0xc0000894f0, 0x1, 0x1, 0x0, 0x0)
| /usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/images/pull.go:160 +0x2d5
| github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).execute(0x5632d26d5e20, 0xc00003c1a0, 0x1, 0x1, 0x5632d26d5e20, 0xc00003c1a0)
| /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:856 +0x47c
| github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x5632d26e7120, 0xc00004a188, 0x5632d1886480, 0x5632d27a98f0)
| /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:974 +0x375
| github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).Execute(...)
| /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:902
| github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).ExecuteContext(...)
| /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:895
| main.Execute()
| /usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/root.go:91 +0xee
| main.main()
| /usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/main.go:39 +0x94
| WARNING: exit code 2 from a shell command.

### Describe the results you expected

image should be download to local disk.
possible to list it using podman images in any shell.

### podman info output

```shell
cannot setresgid: Invalid argument
ERRO[0000] Error getting info on OCI runtime crun: error getting version of OCI runtime crun: /usr/bin/crun --version failed: unknown fcntl argument 1034, assuming long argument.
unknown fcntl argument 1033, assuming long argument.
Failed to re-execute libcrun via memory file descriptor
(exit status 1)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x158 pc=0x562815b020fe]

goroutine 1 [running]:
github.com/containers/podman/libpod.(*Runtime).storeInfo(0xc0004c96c0, 0xc0001f0400, 0x0, 0x0)
/usr/src/packages/BUILD/src/github.com/containers/podman/libpod/info.go:279 +0x7e
github.com/containers/podman/libpod.(*Runtime).info(0xc0004c96c0, 0x0, 0x0, 0x2)
/usr/src/packages/BUILD/src/github.com/containers/podman/libpod/info.go:47 +0x132
github.com/containers/podman/libpod.(*Runtime).Info(...)
/usr/src/packages/BUILD/src/github.com/containers/podman/libpod/runtime.go:909
github.com/containers/podman/pkg/domain/infra/abi.(*ContainerEngine).Info(0xc0000101a8, 0x562816591720, 0xc0001d5140, 0x562815070f4a, 0xc0005ee200, 0xc000147ce8)
/usr/src/packages/BUILD/src/github.com/containers/podman/pkg/domain/infra/abi/system.go:27 +0x34
github.com/containers/podman/cmd/podman/system.info(0x562817137520, 0xc0001120e0, 0x0, 0x1, 0x0, 0x0)
/usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/system/info.go:72 +0x78
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).execute(0x562817137520, 0xc00003c1a0, 0x1, 0x1, 0x562817137520, 0xc00003c1a0)
/usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:856 +0x47c
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x56281713b120, 0xc0000482b8, 0x5628162da480, 0x5628171fd8f0)
/usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:974 +0x375
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).Execute(...)
/usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:902
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).ExecuteContext(...)
/usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:895
main.Execute()
/usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/root.go:91 +0xee
main.main()
/usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/main.go:39 +0x94
```

### Podman in a container

No

### Privileged Or Rootless

None

### Upstream Latest Release

No

### Additional environment details

running on yocto kirkstone.
same recipe was working fine on same buid machine and same podman but on yocto honister version.

### Additional information

Additional information like issue happens only occasionally or issue happens with a particular architecture or on a particular setting

**Comment by rhatdan** [comment]
This is not something that can be fixed upstream, also this is an older version of podman.  Please try with the latest version.

**Reply by Razriel3** [reply]
I have downloaded version 4.3.1 compiled it and installed  it on my ubuntu machine. still having trouble to pull the image.
you say it cannot be fixed upstream.
what do you think is the problem from the error message ? cannot setresgid: Invalid argument

**Reply by mheon** [reply]
Please paste the full error message from 4.3.1

**Comment by rhatdan** [comment]
@giuseppe PTAL

**Comment by giuseppe** [comment]
is podman already running inside a user namespace?  What do you see with `cat /proc/self/uid_map`?

**Comment by Razriel3** [comment]
user@user-Latitude-7410:~/imx-yocto-bsp/sources/meta-coretigo$ cat /proc/self/uid_map
         0          0 4294967295

please notice: I am running with podman 3.4.2
podman pull works well with the honister version of yocto.
podman pull fails with the error messages you see above.

I am attaching devshell for honister with my recipe + devshell for the kirksron with same recipe.

hope this could help.

**Reply by giuseppe** [reply]
thanks.

What do you get with `podman unshare cat /proc/self/uid_map`?

**Comment by Razriel3** [comment]
podman unshare cat /proc/self/uid_map
         0       1000          1
         1     100000      65536

**Reply by giuseppe** [reply]
Thanks.  unshare seems to work fine.

What is the output for `cat /proc/self/status`?

And what do you get with `podman unshare podman pull postgres:14`

**Comment by Razriel3** [comment]
Name:   cat
Umask:  0002
State:  R (running)
Tgid:   3859444
Ngid:   0
Pid:    3859444
PPid:   670358
TracerPid:      0
Uid:    1000    1000    1000    1000
Gid:    1000    1000    1000    1000
FDSize: 256
Groups: 4 24 27 30 46 120 132 133 1000 
NStgid: 3859444
NSpid:  3859444
NSpgid: 3859444
NSsid:  670358
VmPeak:     8232 kB
VmSize:     8232 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:       584 kB
VmRSS:       584 kB
RssAnon:              68 kB
RssFile:             516 kB
RssShmem:              0 kB
VmData:      312 kB
VmStk:       132 kB
VmExe:        20 kB
VmLib:      1652 kB
VmPTE:        52 kB
VmSwap:        0 kB
HugetlbPages:          0 kB
CoreDumping:    0
THP_enabled:    1
Threads:        1
SigQ:   0/62379
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: 0000000000000000
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 000001ffffffffff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Seccomp_filters:        0
Speculation_Store_Bypass:       thread vulnerable
SpeculationIndirectBranch:      conditional enabled
Cpus_allowed:   ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        1
nonvoluntary_ctxt_switches:     0

user@user-Latitude-7410:~/imx-yocto-bsp/sources/meta-coretigo$ podman unshare podman pull postgres:14
Trying to pull docker.io/library/postgres:14...
Getting image source signatures
Copying blob 048d3078d446 done  
Copying blob 76f7157f330d done  
Copying blob 3f4ca61aafcd done  
Copying blob d846f6946dd5 done  
Copying blob c6d23b4fe6c1 done  
Copying blob 4eacfb0464b2 done  
Copying blob 5c197e2b597b done  
Copying blob 2c4576649951 done  
Copying blob 4b8ce16933e9 done  
Copying blob a9ab539fce2b done  
Copying blob 8b507b1498e6 done  
Copying blob 88e4ae5af1fb done  
Copying blob 9a3f79ee6c66 done  
Copying config 277618d300 done  
Writing manifest to image destination
Storing signatures
277618d30078ea526f26522664a116748c53753df87ea093078846666dfc89d5

**Reply by giuseppe** [reply]
so it is working as expected if you run `podman unshare` first?

I don't know about bitbake and what environment it sets up.  So you may want to run these same commands in the environment where podman runs

**Reply by Razriel3** [reply]
from devshell:

podman unshare
cannot setresgid: Invalid argument

podman pull postgres:14
uh-oh!  '/proc/1875468/ns/user' seems to be a symlink, but I can't read it.  Ignoring.uh-oh!  '/proc/1875468/ns/user' seems to be a symlink, but I can't read it.  Ignoring.cannot setresgid: Invalid argument
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x563b891e87a0]

goroutine 1 [running]:
github.com/containers/podman/vendor/github.com/containers/common/libimage.(*Runtime).Pull(0x0, 0x563b8a174720, 0xc0001a6de0, 0xc00010e260, 0xb, 0x0, 0xc00082fa40, 0x0, 0x0, 0x0, ...)
        /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/containers/common/libimage/pull.go:117 +0x500
github.com/containers/podman/pkg/domain/infra/abi.(*ImageEngine).Pull(0xc000612020, 0x563b8a174720, 0xc0001a6de0, 0x7ffcd6ece484, 0xb, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /usr/src/packages/BUILD/src/github.com/containers/podman/pkg/domain/infra/abi/images.go:231 +0x1e5
github.com/containers/podman/cmd/podman/images.imagePull(0x563b8ad0ce20, 0xc000383cc0, 0x1, 0x1, 0x0, 0x0)
        /usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/images/pull.go:160 +0x2d5
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).execute(0x563b8ad0ce20, 0xc00003c1a0, 0x1, 0x1, 0x563b8ad0ce20, 0xc00003c1a0)
        /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:856 +0x47c
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x563b8ad1e120, 0xc0000482b8, 0x563b89ebd480, 0x563b8ade08f0)
        /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:974 +0x375
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).Execute(...)
        /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:902
github.com/containers/podman/vendor/github.com/spf13/cobra.(*Command).ExecuteContext(...)
        /usr/src/packages/BUILD/src/github.com/containers/podman/vendor/github.com/spf13/cobra/command.go:895
main.Execute()
        /usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/root.go:91 +0xee
main.main()
        /usr/src/packages/BUILD/src/github.com/containers/podman/cmd/podman/main.go:39 +0x94

same on honister devshell:

**Reply by giuseppe** [reply]
please show the other commands as well.

What is the output for `cat /proc/self/uid_map` and `cat /proc/self/status`?

**Reply by Razriel3** [reply]
after reboot and running kirkstone devshell

podman unshare
cannot setresgid: Invalid argument
Error: potentially insufficient UIDs or GIDs available in user namespace (requested 0:0 for /home/user/.local/share/containers/storage/overlay/l): Check /etc/subuid and /etc/subgid: chown /home/user/.local/share/containers/storage/overlay/l: invalid argument

cat /proc/self/uid_map
      1000       1000          1

cat /proc/self/status
Name:   cat
Umask:  0022
State:  R (running)
Tgid:   14823
Ngid:   0
Pid:    14823
PPid:   14048
TracerPid:      0
Uid:    1000    1000    1000    1000
Gid:    1000    1000    1000    1000
FDSize: 256
Groups: 65534 65534 65534 65534 65534 65534 65534 65534 1000 
NStgid: 14823
NSpid:  14823
NSpgid: 14823
NSsid:  14046
VmPeak:    99840 kB
VmSize:    99840 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:      2584 kB
VmRSS:      2584 kB
RssAnon:             176 kB
RssFile:            2408 kB
RssShmem:              0 kB
VmData:      432 kB
VmStk:       148 kB
VmExe:        20 kB
VmLib:      1840 kB
VmPTE:        60 kB
VmSwap:        0 kB
HugetlbPages:          0 kB
CoreDumping:    0
THP_enabled:    1
Threads:        1
SigQ:   0/62375
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: 0000000000000000
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 000001ffffffffff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Seccomp_filters:        0
Speculation_Store_Bypass:       thread vulnerable
SpeculationIndirectBranch:      conditional enabled
Cpus_allowed:   ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        3
nonvoluntary_ctxt_switches:     0

**Comment by giuseppe** [comment]
so it is already running in a user namespace where only one ID is mapped:

```
cat /proc/self/uid_map
1000 1000 1
```

Podman won't work in such an environment.  What you could try is to manually create yet another user namespace with `unshare -r` (then run podman from there).  You'll likely also need to set `ignore_chown_errors` in the `storage.conf` file.  Not sure how far it will bring you though.
